mlflow/johnsnowlabs.py

```diff
@@ -0,0 +1,874 @@
+    # be incorrect on multi-node clusters.
+    # If the artifact URI is not a local filesystem path we attempt to write directly to the
+    # artifact repo via Spark. If this fails, we defer to Model.log().
+    elif is_local_uri(run_root_artifact_uri) or not _maybe_save_model(
+            spark_model,
+            append_to_uri_path(run_root_artifact_uri, artifact_path),
+    ):
+        return Model.log(
+            artifact_path=artifact_path,
+            flavor=mlflow.johnsnowlabs,
+            spark_model=spark_model,
+            conda_env=conda_env,
+            code_paths=code_paths,
+            dfs_tmpdir=dfs_tmpdir,
+            sample_input=sample_input,
+            registered_model_name=registered_model_name,
+            signature=signature,
+            input_example=input_example,
+            await_registration_for=await_registration_for,
+            pip_requirements=pip_requirements,
+            extra_pip_requirements=extra_pip_requirements,
+            metadata=metadata,
+        )
+    # Otherwise, override the default model log behavior and save model directly to artifact repo
+    mlflow_model = Model(artifact_path=artifact_path, run_id=run_id)
+    with TempDir() as tmp:
+        tmp_model_metadata_dir = tmp.path()
+        _save_model_metadata(
+            tmp_model_metadata_dir,
+            spark_model,
+            mlflow_model,
+            sample_input,
+            conda_env,
+            code_paths,
+            signature=signature,
+            input_example=input_example,
+            pip_requirements=pip_requirements,
+            extra_pip_requirements=extra_pip_requirements,
+            remote_model_path=remote_model_path,
+        )
+        mlflow.tracking.fluent.log_artifacts(tmp_model_metadata_dir, artifact_path)
+        mlflow.tracking.fluent._record_logged_model(mlflow_model)
+        if registered_model_name is not None:
+            mlflow.register_model(
+                f"runs:/{run_id}/{artifact_path}",
+                registered_model_name,
+                await_registration_for,
+            )
+        return mlflow_model.get_model_info()
+
+
+def _save_model_metadata(
+        dst_dir,
+        spark_model,
+        mlflow_model,
+        sample_input,
+        conda_env,
+        code_paths,
+        signature=None,
+        input_example=None,
+        pip_requirements=None,
+        extra_pip_requirements=None,
+        remote_model_path=None,
+):
+    """
+    Saves model metadata into the passed-in directory.
+    If mlflowdbfs is not used, the persisted metadata assumes that a model can be
+    loaded from a relative path to the metadata file (currently hard-coded to "sparkml").
+    If mlflowdbfs is used, remote_model_path should be provided, and the model needs to
+    be loaded from the remote_model_path.
+    """
+    import pyspark
+
+    if sample_input is not None:
+        mleap.add_to_model(
```

Comment:
since there is a local import of mleap within this called function, `mleap` will need to be a hard-defined dependency of default pip requirements to support this flavor.

Reply:
This is `mlflow.mleap`, I think it's not actually using `mleap `
We can also ask users to add `mleap` to the pip requirements when using `mleap`

Reply:
@C-K-Loan Do we need `mleap`?

Reply:
@harupy  I am not sure to be honest, what are the advantages of supporting mleap?

Reply:
Let's remove it.